[Material] Implement TooltipTheme and Tooltip.textStyle, fix Tooltip debugLabel, update Tooltip defaults

shihaohong opened the pull request:
## Description

1. Implements a TooltipTheme widget that allows all existing properties to be themeable. 
<img width="500" alt="Screen Shot 2019-07-19 at 1 25 50 PM" src="https://user-images.githubusercontent.com/27032613/61563444-cf02ed00-aa28-11e9-984e-faabd7180dd1.png">
<img width="160" alt="Screen Shot 2019-07-19 at 1 25 40 PM" src="https://user-images.githubusercontent.com/27032613/61563428-c6aab200-aa28-11e9-83c1-bd9ab3c9c550.png">

2. Introduces the `textStyle` property, granting more fine-grained control to developers to update the tooltip's text message.
<img width="500" alt="Screen Shot 2019-07-19 at 1 31 22 PM" src="https://user-images.githubusercontent.com/27032613/61563765-91eb2a80-aa29-11e9-9192-45dc403827b7.png">
<img width="160" alt="Screen Shot 2019-07-19 at 1 31 06 PM" src="https://user-images.githubusercontent.com/27032613/61563753-87c92c00-aa29-11e9-8c03-af7e93a24321.png">

3. Fix `debugLabel` to use the correct color when reporting the message's textStyle

4. Breaking change: Updating default tooltip `decoration` to the following: 

- Brightness.light
```
BoxDecoration(
  color: Colors.grey[700].withOpacity(0.9),
  borderRadius: const BorderRadius.all(Radius.circular(4)),
);
```
- Brightness.dark
```
BoxDecoration(
  color: Colors.white.withOpacity(0.9), // new
  borderRadius: const BorderRadius.all(Radius.circular(4)),
);
```
5. Breaking change: Updating default `textStyle` to the following: 

- Brightness.light
```
theme.textTheme.body1.copyWith(
  color: Colors.white, // new
  debugLabel: theme.textTheme.body1.debugLabel.replaceAll('white', 'black'), // fixes debugLabel
);
```
- Brightness.dark
```
theme.textTheme.body1.copyWith(
  color: Colors.black, // new
  debugLabel: theme.textTheme.body1.debugLabel.replaceAll('white', 'black'), // fixes debugLabel
);
```

Tooltip Before Breaking change:
<img width="160" alt="Screen Shot 2019-07-19 at 3 21 15 PM" src="https://user-images.githubusercontent.com/27032613/61568661-f95ca680-aa38-11e9-8c1a-29f59425fbb5.png">
<img width="160" alt="Screen Shot 2019-07-19 at 3 21 26 PM" src="https://user-images.githubusercontent.com/27032613/61568664-fb266a00-aa38-11e9-928d-15775cf4838f.png">

Tooltip After Breaking change: 
<img width="160" alt="Screen Shot 2019-07-19 at 3 19 39 PM" src="https://user-images.githubusercontent.com/27032613/61568668-fd88c400-aa38-11e9-88be-f45ab5360a04.png">
<img width="160" alt="Screen Shot 2019-07-19 at 3 20 08 PM" src="https://user-images.githubusercontent.com/27032613/61568670-feb9f100-aa38-11e9-82ac-5a69aca9b7b8.png">



## Related Issues

Fixes https://github.com/flutter/flutter/issues/35513
Related to https://github.com/flutter/flutter/issues/33654
Related to https://github.com/flutter/flutter/pull/25844

## Tests

I added the following tests:

- Several tests verifying passing down of TooltipTheme properties (color, location, height, decoration, semantics, show and wait durations to reveal and hide tooltip)
- Tests for textStyle implementation
- Test to verify debugLabel for light and dark theme

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

shihaohong committed:
Added TODO comments

shihaohong committed:
Implement TooltipThemeData and TooltipTheme

shihaohong committed:
WIP - Wire TooltipTheme through Tooltip and ThemeData, add TextStyle property

Not documented yet, still needs tests for TooltipTheme.

shihaohong committed:
Add API documentation for textStyle property

shihaohong committed:
Fix wiring of default tooltip values

shihaohong committed:
Add height and Padding tests for TooltipTheme

shihaohong committed:
Add preferBelow and verticalOffset theme tests

shihaohong committed:
Add semantics tests for Tooltip themes

shihaohong committed:
Add tooltip theme waitDuration and showDuration tests

shihaohong committed:
Add custom tooltip TextStyle tests, fix implementation bug

Commit:
Add const where needed in test file

shihaohong committed:
Merge branch 'master' of github.com:flutter/flutter into tooltip-theme

shihaohong committed:
Add custom tooltip TextStyle theme tests

shihaohong committed:
Merge branch 'tooltip-theme' of github.com:shihaohong/flutter into tooltip-theme

shihaohong committed:
Add const to both tests where missing

HansMuller commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,227 @@
+  /// The vertical gap between the widget and the displayed tooltip.
+  final double verticalOffset;
+
+  /// Whether the tooltip defaults to being displayed below the widget.
```
defaults to being displayed below the widget => is displayed below its widget by default

HansMuller commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,227 @@
+  /// tree.
+  final bool excludeFromSemantics;
+
+  /// Specifies the tooltip's shape and background color.
```
Specifies the tooltip's => The tooltip's

HansMuller commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,227 @@
+  /// The amount of time that a pointer must hover over the widget before it
+  /// will show a tooltip.
+  ///
+  /// Defaults to 0 milliseconds (tooltips show immediately upon hover).
```
show => are shown

HansMuller commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,227 @@
+  /// with [ThemeData.typography.white].
+  final TextStyle textStyle;
+
+  /// The amount of time that a pointer must hover over the widget before it
```
The amount of time that a pointer must hover over the widget before it will show a tooltip => The length of time that a pointer must hover over a tooltip's widget before the tooltip will be shown.

HansMuller commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,227 @@
+    assert(t != null);
+    assert(a?.preferBelow == b?.preferBelow);
+    assert(a?.excludeFromSemantics == b?.excludeFromSemantics);
+    if (a == null && b == null)
```
It seems like this statement should be first.

HansMuller commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,227 @@
+    );
+  }
+
+  /// Linearly interpolate between two tooltip themes.
```
If we're going to prohibit changing preferBelow and excludeFromSemantics, we should document that. It might be better to use this kind of "interpolation" instead:
```
      preferBelow: t < 0.5 ? a.preferBelow: b.preferBelow
```

Ideally this stuff wouldn't change but it doesn't seem worth creating a corner case for the odd case where it does.

shihaohong replied:
I'll use that interpolation style instead. It seems to be better than just blocking them outright.

HansMuller commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,227 @@
+       ),
+       super(key: key, child: child);
+
+  /// Specifies the properties for descendant [Tooltip] widgets.
```
Specifies the => The

HansMuller commented on packages/flutter/lib/src/material/tooltip_theme.dart:
````diff
@@ -0,0 +1,227 @@
+  /// Typical usage is as follows:
+  ///
+  /// ```dart
+  /// TooltipTheme theme = TooltipTheme.of(context);
````
```
TooltipThemeData theme = ...
```

HansMuller commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,227 @@
+  /// Specifies the properties for descendant [Tooltip] widgets.
+  final TooltipThemeData data;
+
+  /// The closest instance of this class that encloses the given context.
```
This is the conventional boilerplate for an inherited widget; it's not correct for this theme. This `of` method returns the closest TooltipTheme ancestor's [data] value, and if there is no TooltipTheme ancestor, then it returns the current [Theme]'s tooltipTheme. Apps can assume that the returned value is not null.

HansMuller reviewed (commented):
This all looks good although I'd like to see if we can avoid the Theme brightness dependency. Most of feedback is trivial wordsmithing.

Please add:

DebugFillProperties tests like https://github.com/flutter/flutter/blob/master/packages/flutter/test/material/bottom_sheet_theme_test.dart#L22 and
https://github.com/flutter/flutter/blob/master/packages/flutter/test/material/bottom_sheet_theme_test.dart#L34

A copyWith/hashCode/== test like this: https://github.com/flutter/flutter/blob/master/packages/flutter/test/material/bottom_sheet_theme_test.dart#L10

A test that verifies the theme's default values (even though they're all null; so we don't regress) like this:
https://github.com/flutter/flutter/blob/master/packages/flutter/test/material/button_theme_test.dart#L9

shihaohong commented on packages/flutter/lib/src/material/tooltip.dart:
```diff
@@ -398,33 +417,39 @@ class _TooltipOverlay extends StatelessWidget {
+      ? theme.typography.white.merge(theme.textTheme).body1
+      : theme.typography.white.merge(theme.primaryTextTheme).body1;
+
+    final BoxDecoration defaultDecoration = BoxDecoration(
```
How can this be achieved for the color `Colors.grey[700].withOpacity(0.9)`? I could use `Color(0xFF616161)`, but calling `.withOpacity()` makes it a non-const value.

shihaohong replied:
Actually, after our discussion, if we decide to do a breaking change, the defaults would change as follows: 
```
    final BoxDecoration defaultDecoration = BoxDecoration(
      color: theme.colorScheme.surface,
      borderRadius: const BorderRadius.all(Radius.circular(4)),
    );
```

shihaohong committed:
Incorporate doc improvement and const declaration code review feedback

shihaohong commented on packages/flutter/lib/src/material/tooltip.dart:
```diff
@@ -106,6 +96,16 @@ class Tooltip extends StatefulWidget {
 
+  /// The style to use for the message of the tooltip.
+  ///
+  /// If null, the message's [TextStyle] will be determined based on
```
Should we do the following instead: 
```
    final TextStyle defaultTextStyle = theme.textTheme.body1.copyWith(
      color: theme.colorScheme.onSurface,
    );
```

As discussed offline, it'd be a breaking change, but it'd be closer aligned to the material spec and removes the dependency on Brightness.

shihaohong replied:
Theming is specified by the material spec here: https://material.io/design/components/tooltips.html#theming

shihaohong replied:
I've updated the textTheme to correspond with what is expected by Material

shihaohong committed:
add debugFillProperties tests to tooltip_theme_test.dart

shihaohong committed:
Add tests on TooltipThemeData copyWith, ==, hashCode and default initialization

shihaohong commented:
I've implemented most of the code review suggestions and added the default tests for TooltipThemeData. I'm still currently waiting on what the default values for the border radii, background color, text color and opacity values should be.

shihaohong committed:
Add missing text style and decoration debugFillProperties

shihaohong committed:
Update default colors, add test for debugLabel and dark theme

shihaohong committed:
Updated documentation to reflect updated defaults

shihaohong commented:
I've received feedback on what the default tooltip `decoration` and `textStyle` should be, and this PR is now ready for review.

shihaohong commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,227 @@
+/// Used by [TooltipTheme] to control the properties of tooltips in a widget
+/// subtree.
+///
+/// To obtain the current [TooltipTheme], use [TooltipTheme.of].
```
I updated it to the following:
```
/// To obtain this configuration, use [TooltipTheme.of] to access the closest
/// ancestor [TooltipTheme] of the current [BuildContext].
```

Commit:
Incorporate API documentation feedback

Commit:
Fix grammar on waitDuration doc

shihaohong commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,227 @@
+  final bool preferBelow;
+
+  /// Whether the tooltip's [message] should be excluded from the semantics
+  /// tree.
```
Added an explanation to the API doc

Commit:
Add doc change to Tooltip.preferBelow that reflects change in TooltipTheme

shihaohong committed:
Remove whitespace, add lerp macro doc

shihaohong committed:
Fix properties not properly passing to the _TooltipOverlay

shihaohong committed:
Make sample more fun

HansMuller commented on packages/flutter/lib/src/material/tooltip.dart:
```diff
@@ -92,6 +93,10 @@ class Tooltip extends StatefulWidget {
   /// tree.
+  ///
+  /// Defaults to false. A tooltip will add a [Semantics.label] that is set to
+  /// [Tooltip.message]. Set this property to true if a custom
```
I think you mean ... true if the app is going to provide its own custom semantics label?

shihaohong replied:
Yes! I'll update that

HansMuller commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,251 @@
+  /// tree.
+  ///
+  /// By default, [Tooltip]s will add a [Semantics.label] that is set to
+  /// [Tooltip.message]. Set this property to true if a custom
```
same comment as earlier

HansMuller commented on packages/flutter/lib/src/material/tooltip_theme.dart:
```diff
@@ -0,0 +1,251 @@
+  /// The properties for descendant [Tooltip] widgets.
+  final TooltipThemeData data;
+
+  /// The closest instance of this class' [data] value that encloses the given
```
class' => class's

Maybe this could be a little clearer.
```
Returns the [data] from the closest [TooltipTheme] ancestor.
```

HansMuller commented on packages/flutter/lib/src/material/tooltip.dart:
```diff
@@ -32,55 +34,54 @@ import 'theme_data.dart';
-  /// By default, tooltips prefer to appear below the [child] widget when the
-  /// user long presses on the widget.
+  /// By default, tooltips should adhere to the
+  /// [Material specifications](https://material.io/design/components/tooltips.html#spec).
```
specifications => specification

HansMuller reviewed (approved):
LGTM!

shihaohong committed:
Documentation improvements per feedback by HansMuller

shihaohong committed:
Separate tests to ensure that textStyle is properly reflected

shihaohong committed:
Merge branch 'tooltip-theme' of github.com:shihaohong/flutter into tooltip-theme

shihaohong committed:
Merge branch 'master' of github.com:flutter/flutter into tooltip-theme

shihaohong committed:
Fix tooltipTheme parameter with wrong type

shihaohong committed:
Merge branch 'tooltip-theme' of github.com:shihaohong/flutter into tooltip-theme
